pilot/control/payloads/generic.py

```diff
@@ -148,6 +156,91 @@ def run_payload(self, job, out, err):
                     # also store the full command in case it needs to be restarted later (by the job_monitor() thread)
                     job.utilities[utcmd] = [proc1, 1, utilitycommand]
 
+    def utility_after_payload_finished(self, job):
+        """
+        Functions to run after payload finished
+
+        :param job: job object
+        """
+        log = get_logger(job.jobid, logger)
+
+        # get the payload command from the user specific code
+        pilot_user = os.environ.get('PILOT_USER', 'generic').lower()
+        user = __import__('pilot.user.%s.common' % pilot_user, globals(), locals(), [pilot_user], -1)
+
+        # should any additional commands be prepended to the payload execution string?
+        cmds = user.get_utility_commands_list(order=UTILITY_AFTER_PAYLOAD_FINISHED)
+        if cmds != []:
+            for utcmd in cmds:
+                log.info('utility command to be executed with the payload: %s' % utcmd)
+                # add execution code here
+
+    def pre_payload(self, job):
+        """
+        Functions to run pilot pre payload
+
+        :param job: job object
+        """
+        # write time stamps to pilot timing file
+        add_to_pilot_timing(job.jobid, PILOT_PRE_PAYLOAD, time.time())
+
+    def post_payload(self, job):
+        """
+        Functions to run pilot post payload
+
+        :param job: job object
+        """
+        # write time stamps to pilot timing file
+        add_to_pilot_timing(job.jobid, PILOT_POST_PAYLOAD, time.time())
+
+    def run_payload(self, job, out, err):
+        """
+        (add description)
+
+        :param job: job object
+        :param out: (currently not used; deprecated)
+        :param err: (currently not used; deprecated)
+        :return: proc (subprocess returned by Popen())
+        """
+
+        log = get_logger(job.jobid, logger)
+
+        self.pre_setup(job)
+
+        self.post_setup(job)
```

Comment:
Same comment as before, should be called after setup is done.

Reply:
updated.